tfworker/backends/s3.py

```diff
@@ -116,3 +116,115 @@ def data_hcl(self, remotes: list) -> str:
+            )
+            body = backend_file["Body"]
+            with closing(backend_file["Body"]):
+                backend = json.load(body)
+
+            if validate_backend_empty(backend):
+                self.delete_with_versions(s3_object)
+                click.secho(f"backend file removed: {s3_object}", fg="yellow")
+            else:
+                raise BackendError(f"backend at: {s3_object} is not empty!")
+
+    def _clean_locking_state(self, deployment, definition=None):
+        """
+        clean_locking_state when called removes the dynamodb table
+        that holds all of the state checksums and locking table
+        entries
+        """
+        dynamo_client = self._authenticator.backend_session.resource("dynamodb")
+        if definition is None:
+            table = dynamo_client.Table(f"terraform-{deployment}")
+            table.delete()
+            click.secho(f"locking table: terraform-{deployment} removed", fg="yellow")
+        else:
+            # delete only the entry for a single state resource
+            table = dynamo_client.Table(f"terraform-{deployment}")
+            table.delete_item(
+                Key={
+                    "LockID": f"{self._authenticator.bucket}/{self._authenticator.prefix}/{definition}/terraform.tfstate-md5"
+                }
+            )
+            click.secho(
+                f"locking table key: '{self._authenticator.bucket}/{self._authenticator.prefix}/{definition}/terraform.tfstate-md5' removed",
+                fg="yellow",
+            )
+
+    def clean(self, deployment: str, limit: list = None) -> None:
+        if limit:
+            for limit_item in limit:
+                click.secho(
+                    "when using limit, dynamodb tables won't be completely dropped",
+                    fg="yellow",
+                )
+                try:
+                    # the bucket state deployment is part of the s3 prefix
+                    self._clean_bucket_state(definition=limit_item)
+                    # deployment name needs specified to determine the dynamo table
+                    self._clean_locking_state(deployment, definition=limit_item)
+                except BackendError as e:
+                    click.secho(f"error deleting state: {e}", fg="red")
+                    raise SystemExit(1)
+        else:
+            try:
+                self._clean_bucket_state()
+            except BackendError as e:
+                click.secho(f"error deleting state: {e}")
+                raise SystemExit(1)
+            self._clean_locking_state(deployment)
+
+    def delete_with_versions(self, key):
+        """
+        delete_with_versions should handle object deletions, and all references / versions of the object
+
+        note: in initial testing this isn't required, but is inconsistent with how S3 delete markers, and the boto
+        delete object call work there may be some selfurations that require extra handling.
```

Comment:
Is `selfurations` supposed to be `configurations`?

Reply:
probably, updated!